### DISCUSSION TITLE
Art Assets

### DISCUSSION THREAD

**Discussion by BaconCommander** [Thorium-Sim/thorium-nova#11] [Ideas]
Figuring out how to handle assets is something going on in my head; the idea was put forward to have a seperate repository just for assets, which works for me. I'm also hoping to figure out a good best practices for assets; Image sizes, Polycounts, Etc.

**Comment by alexanderson1993** [comment]
A couple of suggestions:
- We could store assets within the Thorium Nova repo using Git LFS, but we would be limited to 1GB of storage and 1GB of download bandwidth without paying.
- We could store assets in a separate repo, which makes it easier to access the code without needing to manage all of the assets within the same git repo. This seems a good option to me.
- We could use a third-party file storage solution, like Dropbox or Google Drive. I would prefer the latter.
- We could roll our own solution using commodity file storage services, like S3 or Backblaze B2 (which is what I already use.)

Anyone have any other thoughts?

**Comment by alexanderson1993** [comment]
When it comes to best practices for assets, there are three key things to remember:
- We're running these in a browser, so we have to deal with those limitations.
- We don't know what kinds of hardware this will run on, so CPU, GPU, and memory limits have to be factored in.
- Probably the biggest constraint: All of the assets have to be downloaded over the network quickly, so there isn't a long "Loading..." screen.

Much of the asset downloading can be hidden with code with smart caching and slight-of-hand, but keeping asset sizes small is still a good practice.

**Comment by BaconCommander** [comment]
Yeah, putting in proper limitations for assets will be rather crucial. I like both the ideas of having a separate git hub repo or using google drive to store the assets.  Both have appeal to me.  Between the two I feel google drive might be the simpler solution.

**Reply by alexanderson1993** [reply]
Agreed with Google Drive. Lets wait a few more hours for more suggestions to come in, and if there are none I'll get a Google Drive set up and linked here.

**Reply by astrolamb-gaming** [reply]
I'm not opposed to Google Drive due to its simplicity for interaction by anyone, regardless of their knowledge of git. However, using Google Drive does add another aspect of the group to manage; it's another platform where contributors have to be added to access files and there is no reverting if someone accidentally deletes a file. In that respect using another github repo would be better, to keep it all as much together as we can, and, of course, as source control. 
I'm not quite sure how access to files on GD would work for the user. Would all files be accessible in a read-only manner by everyone? Is there an extra api or somesuch that would be required? Would the assets be manually/automatically added to the release files?

**Reply by alexanderson1993** [reply]
All good points. I'll have to ponder on it a little bit more. I think this is worthy of a vote, once we've had a bit more discussion.

There's a distinction to be made, too - Development assets, like full-res textures, .blend files, .psd, etc; and production assets, like the .glb models, compressed images, etc. I'm certain some kind of automation can be set up to pull the correct assets from Google Drive using their API, but that's a good bit of work that might be easier for a Github repo.

**Reply by alexanderson1993** [reply]
Let me throw one more tidbit in here - the assets for Thorium Classic are a total joke. It's just a .zip file that I store on S3 that gets downloaded in the background the first time you run Thorium. I never update it or add new assets, so it's really stagnant that way.

Come to think of it, I kind of touched on this in the post on Plugins from last year (https://thoriumsim.com/blog/plugins-how-to-configure-thorium-nova). The way I had it set up with Thorium Nova Prime (I think that's what I'll call the old version of Thorium Nova), all assets were bundled into plugins. There is one plugin which ships with Thorium Nova that has the base game, including missions, ship definitions, assets, etc. 

So ultimately this becomes a broader discussion of the plugin authoring system and experience. 

@BaconCommander Any idea how big the source files are going to be? It think that is going to factor into the decision as to whether we use Github or some other storage platform.

**Reply by astrolamb-gaming** [reply]
I actually have written a couple websites that pull data and/or new code from github. It's easy to check if there's been a new release, and if so, to pull the new stuff for use. In those cases it just pulls the whole repo instead of assets that don't already exists (which an actual git intergration somehow would do), but it's pretty convenient and doesn't need any authorization. Not sure if that's a route we want to go though.

I like the idea of using plugins as a way to implement both base game assets as well as any custom assets. I agree with Xansta's post below that it should be simple for "modders" and contributors to add new things.

**Reply by alexanderson1993** [reply]
> I like the idea of using plugins as a way to implement both base game assets as well as any custom assets. 

That's my plan as well, see the [discussion I made earlier today](https://github.com/Thorium-Sim/thorium-nova/discussions/26). 

This then becomes two decisions:

- Where do we store and collaborate on source/development assets (.blend, .psd, etc)?
- Where do we store and collaborate on plugins (which include the built/production assets)?

For the latter, if we were using Github (which I'm starting to think that we should) there would need to be some way to make all of that work with Thorium's built-in plugin editor. Or maybe the version control and collaboration aspect should be built into Thorium outright (although that is a lot of work)

What's the likelihood that anyone other than direct Thorium contributors would want to collaborate on their plugins with other people?

**Comment by Xansta** [comment]
Whatever gets decided, I suggest it be set up to make it easy for anyone to include their own artwork for a specific mission and make it easy for anyone to contribute artwork to the overall project (two different use cases)
